[FIXED JENKINS-30502] trim job name upon rename

csimons opened the pull request:
If a job is renamed to a string with trailing whitespace, exceptions are thrown as the application attempts to rename the job directory in the file system, and the job can be neither deleted nor renamed again through the UI. This problem occurs only when renaming a job, and not when initially creating a job, because job names are trimmed when first created.  This change trims the job name for rename operations as well, so that the name-setting behavior is consistent across create/update operations and inadvertent leading/trailing whitespace is removed before it can cause problems.

oleg-nenashev commented:
:+1: 
In order to improve the PR, I would recommend using brackets even for single-liners. A simple test in test-harness suite would be also useful

KostyaSha commented:
This was discussed many times AFAIR cc @daniel-beck 
AFAIR the idea was allow having whitespace during rename, but strip for newly created jobs because there was no strip before and that's why rename was keeped.

KostyaSha commented:
Short search on jenkins issues https://issues.jenkins-ci.org/browse/JENKINS-19203

KostyaSha commented on core/src/main/java/hudson/model/Job.java:
```diff
@@ -1209,6 +1209,9 @@ public synchronized void doConfigSubmit(StaplerRequest req,
             ItemListener.fireOnUpdated(this);
 
             String newName = req.getParameter("name");
+            if (newName != null)
```
```
if () {
  code
}
```

csimons commented:
@KostyaSha @oleg-nenashev 

(1) I've added the if-block braces.

(2) I have updated the logic, in ff8668168eb5929cb91fca4400630f8a66c8d86a, to trim the new name only if it differs from the old name.  This ensures job names aren't automatically renamed without the user's knowledge, as @daniel-beck pointed out would be undesirable during the JENKINS-19203 discussion, but instead are trimmed only upon new rename operations.

KostyaSha commented on core/src/main/java/hudson/model/Job.java:
```diff
@@ -1211,6 +1211,7 @@ public synchronized void doConfigSubmit(StaplerRequest req,
             String newName = req.getParameter("name");
             final ProjectNamingStrategy namingStrategy = Jenkins.getInstance().getProjectNamingStrategy();
             if (newName != null && !newName.equals(name)) {
+                newName = newName.trim();
```
Wouldn't this break equals matching?

KostyaSha commented:
Keeping it for @daniel-beck

daniel-beck commented:
Looks like a pragmatic approach, so :+1: from me.

KostyaSha commented:
i trust @daniel-beck :+1:

oleg-nenashev commented:
I have a little concern...
You check <code>newName.equals(name)</code> before trimming, so the rename may happen if I submit "Foo " for the existing "Foo". I suppose such case should be ignored via the double check.

Commit:
[FIXED JENKINS-30502] trim job name upon rename

If a job is renamed to a string with trailing whitespace, exceptions are
thrown as the application attempts to rename the job directory in the
file system, and the job can be neither deleted nor renamed again
through the UI. This problem occurs only when renaming a job, and not
when initially creating a job, because job names are trimmed when first
created.  This change trims the job name for rename operations as well,
so that the name-setting behavior is consistent across create/update
operations and inadvertent leading/trailing whitespace is removed before
it can cause problems.

csimons commented:
@oleg-nenashev Good catch.  I've updated the .equals() check to be more robust and handle this case properly.

KostyaSha commented on core/src/main/java/hudson/model/Job.java:
```diff
@@ -1210,7 +1210,8 @@ public synchronized void doConfigSubmit(StaplerRequest req,
             String newName = req.getParameter("name");
             final ProjectNamingStrategy namingStrategy = Jenkins.getInstance().getProjectNamingStrategy();
-            if (newName != null && !newName.equals(name)) {
+            if (newName != null && !newName.trim().equals(name.trim())) {
```
looks ugly :(

kzantow replied:
@KostyaSha perhaps you have a suggestion to make it 'prettier'? Rather than simply noting the fact you don't like something, providing a way to improve it would more easily result in it getting 'fixed'. For example, do you know of a utility method in Jenkins that would make this easier to understand?

KostyaSha replied:
@kzantow it just a marker, probably somebody else will suggest something better.

KostyaSha replied:
I guess newName shouldn't be also empty, so commons or guava StringUtils with check nonEmptyNull should work, Then trim, then make comparison?

KostyaSha replied:
@kzantow as i see you are newly core contributor, do you have any argument about block of existed code i.e. "no, it works wonderful", "you are wrong, this is good logic" rather then discussing my comment?
@kzantow as core contributor please suggest something for this block of code.

oleg-nenashev replied:
@KostyaSha 
As a **core contributor** with pretty **much contributions** I think your comments are totally invalid:
* the initial comment didn't contain any feedback regarding the code, hence @kzantow 's comment was totally valid
* your offensiveness is not constructive here

I also think the code is good enough

kzantow replied:
@KostyaSha my comment was a suggestion for you to provide insight rather than something which offers no solution. I also agree with @oleg-nenashev , the code seems fine (there's a null check on the incoming data; assume preexisting 'name' is not null), adding anything more will only help to confuse what it's doing.

KostyaSha replied:
> the initial comment didn't contain any feedback regarding the code, hence @kzantow 's comment was totally valid

@oleg-nenashev i following logical changes in this PR and said what i think, if it not enough for you, then just ask me without pointing me that i should expect from somebody full understanding.  

> your offensiveness is not constructive here

I treat as offensiveness comments in my side. If many trim operations in two lines of code is ok for you and other people likes it, then i would not mind. But i don't see any feedback to this change or PR from this commenter. 
:+1: and :-1: exists for getting quorum to get better quality.

lanwen replied:
Thinks it can be writed with help of StringUtils from commons-lang

with static import it can be:
```java
import static org.apache.commons.lang.StringUtils.trim;

//...
if(!StringUtils.equals(trim(name), trim(newName))) {
    newName = trim(newName);
}
```

KostyaSha replied:
It comment that i expected, at least here code logic is better readable.
@csimons WDYT?

lanwen replied:
if `name` can be null it should be `trimToEmpty(name)` instead

csimons replied:
@KostyaSha In response to your WDTY query, I understand and agree with the sentiment that in general it's not great to have a bunch of trim()s everywhere.  In this case, however, the bug fix is all about trimming and checking the trimming carefully, and I'm afraid we aren't going to get around having to actually trim these strings. :)  I like the suggestions about trying to combine stuff using utility methods, but the !StringUtils.equals(trim(name), trim(newName)) suggested above is actually longer than !newName.trim().equals(name.trim()), so the code as-is seems about as good as we'll be able to get as far as I can see.  If anyone has a better solution that maintains _exactly_ the logic as it is here (which is required to resolve this bug) I'm absolutely willing to update it.

lanwen replied:
@csimons you should compare whole line 
`newName != null && !newName.trim().equals(name.trim())` with 
`!StringUtils.equals(trim(name), trim(newName)) ` which actually shorter, even in form
`!StringUtils.equals(trimToEmpty(name), trim(newName))`

csimons replied:
@lanwen The whole line,

```
newName != null && !newName.trim().equals(name.trim())
```

and 

```
!StringUtils.equals(trim(name), trim(newName))
```

are _not_ logically equivalent.  In the latter case, if newName is null, the body of the if-block will be entered.  In the former case, the block will _not_ be entered..

csimons replied:
@lanwen The original line,
```
newName != null && !newName.trim().equals(name.trim())
```
and
```
newName != null && !newName.trim().equals(name.trim())
```
are not logically equivalent.  In the former case, a null newName will result in _not_ entering the body of the if-block, but in the latter case, it would be entered under this condition.

If we use the StringUtils methods here, with the static import of #trim, the complete expression will be
```
(newName != null && !StringUtils.equals(trim(newName), trim(name)))
```
which is more verbose than the original
```
(newName != null && !newName.trim().equals(name.trim())
```
So the line would be longer and we'll need to include two additional import statements.  The original verbiage seems most concise.

KostyaSha commented:
Oh, btw, no test cases. :-1: until mentioned behaviour wouldn't be catched with tests.

kzantow commented:
-1 agree with @KostyaSha a simple test case would be worthwhile

csimons commented:
In e7790b9fa608d34d93ed8ab281d736759041f817 I have included a test case, with two trivial cases implemented as below:

```
@Issue("JENKINS-30502")
@Test public void testRenameTrimsSpace() throws Exception {
    tryRename("myJob", " foo", "foo");
    tryRename("myJob", "foo ", "foo");
}
```

I'd like to include more tests to protect against the issues @daniel-beck mentioned in JENKINS-19203, as below, but unfortunately it seems there's no good way to get these tests working on Windows machines as Windows does not allow trailing space in directory names, and so creating these jobs on Windows will throw IOException:[1][2]

```
@Issue("JENKINS-30502")
@Test public void testRenameTrimsSpace() throws Exception {
    // Don't auto-trim existing untrimmed names (disruptive).
    tryRename("myJob ", "myJob ", "myJob ");

    // Allow for removing untrimmed whitespace by user.
    tryRename("myJob ", "myJob", "myJob");

    // Trim new renames regardless of (un)trimmed prior name.
    tryRename("myJob", " foo", "foo");
    tryRename("myJob", "foo ", "foo");
    tryRename("myJob ", " foo", "foo");
    tryRename("myJob ", "foo ", "foo");
    tryRename(" myJob", " foo", "foo");
    tryRename(" myJob", "foo ", "foo");
}
```

I could write some conditional code to detect the OS and then do these tests only if we're on a non-Windows OS, and am willing to do this if we're open to making the test suite behave differently when executed on different platforms, but I'm not sure what the project policy about this is (I'd love to know).

[1] https://bugs.openjdk.java.net/browse/JDK-6578565
[2] https://msdn.microsoft.com/en-US/library/aa365247.aspx

KostyaSha commented on test/src/test/java/hudson/model/JobTest.java:
```diff
@@ -286,4 +291,35 @@ private static void tryConfigDotXml(JenkinsRule.WebClient wc, int status, String
+
+        String urlTemplate = "/job/{0}/rename?newName={1}";
+        String expectedURL = MessageFormat.format(
+                urlTemplate, initialName, correctResult);
```
Doesn't this fit into 120 width?

csimons replied:
Yes.  I tried to keep in under 80 characters as I thought I read somewhere that the Jenkins project that we were loosely following the Sun/Oracle standards (which is 80 characters[1]), but if the project policy is 120 I can update this.  Are code style guidelines for this project published anywhere?

[1] http://www.oracle.com/technetwork/java/javase/documentation/codeconventions-136091.html

KostyaSha replied:
java never fits into 80, use 120.

csimons replied:
The "Archival Purposes Only" notice doesn't change that the posted guidelines at the Jenkins website states we're using this standard.[1]

[1] https://wiki.jenkins-ci.org/display/JENKINS/contributing

KostyaSha commented on test/src/test/java/hudson/model/JobTest.java:
```diff
@@ -286,4 +291,35 @@ private static void tryConfigDotXml(JenkinsRule.WebClient wc, int status, String
+        assertTrue(resultPage.getUrl().toString().endsWith(expectedURL));
+    }
+
+    private FreeStyleProject createJobIfNeeded(String jobName)
```
No need in checking for existed jobs, just name them unique. And `@Rule` runs fresh jenkins instance for any new `@Test`

lanwen replied:
yep, should be no dependencies between tests
:-1:

csimons replied:
I have made the job names unique and have made all tests independent.

KostyaSha commented on test/src/test/java/hudson/model/JobTest.java:
```diff
@@ -286,4 +291,35 @@ private static void tryConfigDotXml(JenkinsRule.WebClient wc, int status, String
+        String expectedURL = MessageFormat.format(
+                urlTemplate, initialName, correctResult);
+
+        assertTrue(resultPage.getUrl().toString().endsWith(expectedURL));
```
add some human readable message either it will be difficult understand error message from test suite output.
Or even better use https://code.google.com/p/hamcrest/wiki/Tutorial `assertThat()`

lanwen replied:
`assertThat(resultPage.getUrl().toString(), endsWith(expectedURL)` returns full debug instead of just "false when true expected"

KostyaSha commented on test/src/test/java/hudson/model/JobTest.java:
```diff
@@ -286,4 +291,35 @@ private static void tryConfigDotXml(JenkinsRule.WebClient wc, int status, String
+    }
+
+    private void tryRename(
+            String initialName,
```
strange style, looks like you trying to fit into calculator width :)
Please place first argument after `(`

csimons replied:
Done.

KostyaSha commented:
> (I'd love to know).

If test sensitive to OS's then it welcome to make tests on both platforms. 
As example of skipping https://github.com/jenkinsci/jenkins/pull/1845

lanwen commented on test/src/test/java/hudson/model/JobTest.java:
```diff
@@ -286,4 +291,35 @@ private static void tryConfigDotXml(JenkinsRule.WebClient wc, int status, String
 
+    @Issue("JENKINS-30502")
+    @Test public void testRenameTrimsSpace() throws Exception {
+        tryRename("myJob", " foo", "foo");
```
please use Parametrization instead of actually 2 cases in one method, or split it to separate methods, as of when fails first sentence, the second will never be run.

KostyaSha replied:
This can degrade test time, and in some kind of logic this test relates to one piece of logic

lanwen replied:
this makes dependency between tests and makes author to create methods like `createJobIfNeeded`

csimons replied:
This has been done in the latest revision.

KostyaSha commented on test/src/test/java/hudson/model/JobTest.java:
```diff
@@ -286,4 +291,35 @@ private static void tryConfigDotXml(JenkinsRule.WebClient wc, int status, String
     }
 
+    @Issue("JENKINS-30502")
+    @Test public void testRenameTrimsSpace() throws Exception {
```
Please put annotation on new line, coding style that you see is inventing by one guy and has bad readability :)

csimons replied:
Yep, I was copying the style from earlier in this file.  In the latest revision this has been adjusted.

KostyaSha commented on core/src/main/java/hudson/model/Job.java:
```diff
@@ -1210,7 +1212,8 @@ public synchronized void doConfigSubmit(StaplerRequest req,
             String newName = req.getParameter("name");
             final ProjectNamingStrategy namingStrategy = Jenkins.getInstance().getProjectNamingStrategy();
-            if (newName != null && !newName.equals(name)) {
+            if (!StringUtils.equals(trim(newName), trim(name))) {
```
does `trim()` do null check? @lanwen mentioned about `trimNotNull`

csimons replied:
trim() returns null for null input, and StringUtils.equals() handles nulls as desired here.  I can change this to trimNotNull, but I thought the point of adding the StringUtils methods here in the first place was to make the line more concise.

KostyaSha commented on core/src/main/java/hudson/model/Job.java:
```diff
@@ -72,6 +72,8 @@
 
 import org.apache.commons.io.FileUtils;
+import org.apache.commons.lang.StringUtils;
+import static org.apache.commons.lang.StringUtils.trim;
```
static imports usually imported after all imports. Do you manage it manually? Usually all java IDEs handling this.

csimons replied:
This has been changed as requested.

csimons commented:
@KostyaSha Thanks for the reference about OS-dependent tests in #1845.

As of cead12e033d1a3e9942422c97e886fdb46e0823c I have included the OS-dependent tests for this change.

csimons commented:
After getting back home to a Linux machine I was able to run the full test suite and found that the initial logic wasn't sophisticated enough after all, so I've pulled the if-block condition out into a separate method for clarity, and made the test cases a bit more clear.  I think all suggestions have been incorporated and this is ready to be reviewed (assuming the build succeeds).

oleg-nenashev commented:
Looks good to me. :+1:

KostyaSha commented:
:+1:  lgtm

lanwen commented:
lgtm
